misc/python/materialize/mzbuild.py

```diff
@@ -363,36 +451,82 @@ def prepare_batch(cls, cargo_builds: list["PreImage"]) -> str:
         builds = cast(list[CargoBuild], cargo_builds)
         bins = set()
         examples = set()
+        bazel_bins = dict()
+        bazel_tars = dict()
         for build in builds:
             if not rd:
                 rd = build.rd
             bins.update(build.bins)
             examples.update(build.examples)
+            bazel_bins.update(build.bazel_bins)
+            bazel_tars.update(build.bazel_tars)
         assert rd
 
         ui.section(f"Common cargo build for: {', '.join(bins | examples)}")
-        cargo_build = cls.generate_cargo_build_command(rd, list(bins), list(examples))
+
+        if rd.bazel:
+            cargo_build = cls.generate_bazel_build_command(
+                rd, list(bins), list(examples), bazel_bins, bazel_tars
+            )
+        else:
+            cargo_build = cls.generate_cargo_build_command(
+                rd, list(bins), list(examples)
+            )
+
         spawn.runv(cargo_build, cwd=rd.root)
 
         # Re-run with JSON-formatted messages and capture the output so we can
         # later analyze the build artifacts in `run`. This should be nearly
         # instantaneous since we just compiled above with the same crates and
         # features. (We don't want to do the compile above with JSON-formatted
         # messages because it wouldn't be human readable.)
-        json_output = spawn.capture(
-            cargo_build + ["--message-format=json"],
-            cwd=rd.root,
-        )
-
-        return json_output
-
-    def build(self, cargo_build_json_output: str) -> None:
+        if not rd.bazel:
+            json_output = spawn.capture(
+                cargo_build + ["--message-format=json"],
+                cwd=rd.root,
+            )
+            prep = {"cargo": json_output}
+        else:
+            # TODO(parkmycar): Having to assign the same compilation flags as the build process
+            # is a bit brittle. It would be better if the Bazel build process itself could
+            # output the file to a known location.
```

Comment:
Agreed, but no worse than what we currently do for Cargo, where we invoke twice to get human readable output the first time and machine readable output the second time, and we really need that second invocation to be identical to the first to avoid recompiling everything.

Reply:
Sounds good to me!